streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,638 @@
+        if world.is_local_leader:
+            sequences = get_epoch(self.shard_sizes, self.shuffle, self.seed, epoch, sessions)
+            base = world.node * world.ranks_per_node * world.workers_per_rank
+            for rank_of_node in range(world.ranks_per_node):
+                for worker_of_rank in range(world.workers_per_rank):
+                    worker = base + rank_of_node * world.workers_per_rank + worker_of_rank
+                    name = f'{self._prefix}_part_{worker:03}'
+                    sequence = sequences[worker]
+                    size = len(sequence) * np.int64(0).nbytes
+                    if not size:
+                        continue
+                    shm = SharedMemory(name, True, size)
+                    shm.buf[:] = sequence.tobytes()
+
+        self._barrier()
+
+        # Load our partition.
+        name = f'{self._prefix}_part_{world.worker:03}'
+        try:
+            shm = SharedMemory(name)
+        except:
+            return None
+        todos = np.frombuffer(shm.buf, np.int64).copy()
+        shm.unlink()
+
+        return todos
+
+    def _download_file(self, basename: str) -> str:
+        """Safely download a file from remote to local cache.
+
+        Args:
+            basename (str): Basename of file to download.
+
+        Returns:
+            str: Local cache filename.
+        """
+        if self.remote is None:
+            remote = None
+        else:
+            remote = os.path.join(self.remote, self.split, basename)
+        local = os.path.join(self.local, self.split, basename)
+        for _ in range(1 + self.download_retry):
+            try:
+                download(remote, local, self.download_timeout)
+            except:
+                continue
+            break
+        return local
+
+    def _decompress_shard_part(self, zip_info: FileInfo, zip_filename: str, raw_filename: str,
+                               compression: Optional[str]) -> None:
+        """Validate and decompress shard data.
+
+        Args:
+            zip_info (FileInfo): Compressed file info.
+            zip_filename (str): Compressed filename.
+            raw_filename (str): Decompressed filename.
+            compression (str, optional): Compression algorithm.
+        """
+        # Load compressed.
+        data = open(zip_filename, 'rb').read()
+
+        # Validate what was downloaded.
+        if self.validate_hash:
+            assert get_hash(self.validate_hash, data) == zip_info.hashes[self.validate_hash]
+
+        # Decompress and save that.
+        data = decompress(compression, data)  # pyright: ignore
+        tmp_filename = raw_filename + '.tmp'
+        with open(tmp_filename, 'wb') as out:
+            out.write(data)
+        os.rename(tmp_filename, raw_filename)
+
+        # Maybe remove compressed to save space.
+        if not self.keep_zip:
+            os.remove(zip_filename)
+
+    def _download_shard_part(self,
+                             shard: int,
```

Comment:
looks like `shard` variable is not getting used in this function !!!!

Reply:
Done